modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistryImpl.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.metric;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+import java.util.function.DoubleSupplier;
+import java.util.function.IntSupplier;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.processors.metric.impl.BooleanGauge;
+import org.apache.ignite.internal.processors.metric.impl.DoubleGauge;
+import org.apache.ignite.internal.processors.metric.impl.IntGauge;
+import org.apache.ignite.internal.processors.metric.impl.LongGauge;
+import org.apache.ignite.internal.processors.metric.impl.ObjectGauge;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.spi.metric.Metric;
+import org.apache.ignite.internal.processors.metric.impl.DoubleMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.HitRateMetric;
+import org.apache.ignite.internal.processors.metric.impl.IntMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.LongAdderMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.LongMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.BooleanMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.HistogramMetric;
+import org.apache.ignite.internal.processors.metric.impl.ObjectMetricImpl;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.processors.metric.impl.MetricUtils.metricName;
+import static org.apache.ignite.internal.util.lang.GridFunc.nonThrowableSupplier;
+
+/**
+ * Simple implementation.
+ */
+public class MetricRegistryImpl implements MetricRegistry {
+    /** Registered metrics. */
+    private final ConcurrentHashMap<String, Metric> metrics = new ConcurrentHashMap<>();
```

Comment:
Each key here represents a full hierarchy string and likely to consume a lot of heap space.
Please provide a test measuring heap occupation with large number of registered metrics.
You could look at GridAffinityAssignmentJolBenchmark for reference.
Probably better to use some tree-like strucure here (trie)

Reply:
1. `GridMetricsJolBenchmark` added.

2. I assume we should optimize the internal implementation of `MetricRegistry` for frequently reading. Seems, it is opposite to lower memory consumption. What do you think?

3. Anyway, I will create a ticket to spend some extra time to optimize the internal implementation of `MetricsRegistry` after all current Ignite metrics will be migrated to it.

Reply:
[IGNITE-11912](https://issues.apache.org/jira/browse/IGNITE-11912) created.

Reply:
Metrics are not expected to be read by key oftenly from my view. 
Which scenarios do you have in mind? 
Find for me to address this in separate ticket.

Reply:
> Metrics are not expected to be read by key

Each exporter iterate through the set(or subset) of metrics and essentially write in some specific format following information:

1. Metric name.
2. Metric value.
3. Optionally, metric description.

If this scenario is basic then we should keep metric name ready to be written(or returned) somewhere.

Anyway, I agree with you that we should optimize internal implementation. I will investigate this issue in IGNITE-11912.

Thanks for pointing this out.